### PR TITLE
feat: implement activities CRUD with many-to-many categories and pagi…

### PR DESCRIPTION
**Pull request by hill-ol**
## Description

This PR implements the complete activities feature for trip planning, allowing users to propose, categorize, and manage activities within trips. The implementation uses a many-to-many relationship between activities and categories, enabling flexible organization where a single activity (e.g., "Dinner and Show") can belong to multiple categories (e.g., "food" and "entertainment").

### High-Level Approach

The feature follows the established three-layer architecture pattern (Repository → Service → Controller):

**Database Layer:**
- `categories` table stores trip-scoped categories with composite PK (trip_id, name)
- `activities` table stores activity details including dates as JSONB arrays
- `activity_categories` join table implements the many-to-many relationship

**Repository Layer:**
- Activity repository handles CRUD with cursor pagination and batch category fetching for performance
- Category repository manages trip-scoped categories
- ActivityCategory repository manages the many-to-many relationships with idempotent operations

**Service Layer:**
- Business logic includes permission validation (trip membership required)
- Automatic category creation when referenced in activities
- Transactional operations to ensure atomicity (e.g., activity + categories created together)
- Batch fetching of categories for multiple activities to avoid N+1 queries

**Controller Layer:**
- RESTful endpoints following team conventions
- Request validation and error handling
- Swagger documentation for all endpoints

### Key Design Decisions

**Many-to-Many Relationship:** Activities can have multiple categories, implemented via join table rather than an array field. This allows for efficient querying by category and maintains referential integrity.

**Cursor-Based Pagination:** All list endpoints use cursor pagination (created_at, id) for consistent ordering and performance at scale, following the pattern established in memberships.

**Default Categories:** Five default categories (food, lodging, attraction, transportation, entertainment) are automatically created when a trip is initialized, executed within the same transaction as trip creation for atomicity.

**Dates as JSONB:** Activity dates stored as JSONB array of date ranges, allowing multiple date options and single API call updates as specified in the ticket.

**Category Auto-Creation:** When an activity references a non-existent category, it's automatically created for that trip, reducing friction in the user experience.

### How It Fits Into the System

Activities integrate with existing features:
- **Trips:** Activities are scoped to trips, cascade delete when trip is deleted
- **Memberships:** Permission checking uses existing membership validation
- **Users:** Activity proposer linked to users, includes profile picture integration
- **Comments:** Activities can be commented on (using existing comment system with ActivityEntity type)
- **File Service:** Profile pictures for activity proposers use existing S3 integration

### Integration Tests (30 tests, all passing)

**TestActivityLifecycle (13 tests):**
- Create activity with multiple categories
- Get single activity with all details
- List activities for a trip
- Update activity fields (name, description, dates)
- Delete as proposer and as admin
- Permission checks (non-admin member cannot delete others' activities)
- Non-member access prevention (create, view, list)
- Custom category auto-creation
- Error cases (not found, invalid UUIDs)

**TestActivityCategories (7 tests):**
- Get paginated categories for an activity
- Add single category to activity
- Add multiple categories to activity
- Remove category from activity
- Non-member permission checks for category operations
- Idempotency of adding same category twice

**TestActivityPagination (6 tests):**
- Pagination with multiple pages and cursor navigation
- Edge cases (zero limit, invalid cursor)
- Empty result handling
- Default limit behavior (20 items)
- No duplicates across pages
- Correct ordering (created_at DESC, id DESC)

**TestCategoryGetByTripID (4 tests):**
- Member can get trip categories
- Non-member access denied
- Default categories created on trip initialization
- Custom categories visible after activities created

### Manual Testing Plan
- Tested migrations locally with rollback
- Verified cascade deletes (trip deletion removes activities and categories)
- Confirmed transaction rollback on errors
- Validated cursor pagination with large datasets

## Checklist

* [x] I have self-reviewed my code for readability, maintainability, performance, and added comments/documentation where necessary.
  
* [x] New and existing tests pass locally with my changes.
  
* [x] I have followed the project's coding standards and best practices.

- User-visible changes:
  - Full activity management: create, read, update, delete activities under a trip with cursor-based pagination; responses include proposer info and resolved proposer profile picture.
  - Activity↔Category many-to-many: add/remove/list categories per activity, list trip-scoped categories, automatic creation of referenced categories, and five default categories created transactionally with trip creation.
  - Dates as JSONB arrays: activities support multiple date ranges via DateRange arrays.
  - Permissioned, transactional behavior: membership-based permission checks, idempotent category linking, and transactional creation/update/removal flows.
  - API surface & docs expanded: Swagger/OpenAPI adds Activity models, Create/Update requests, DateRange, paginated responses, and category endpoints; controllers, routers, and validators wired and documented.

- Categorized change list:
  - Features
    - Activity CRUD with cursor pagination and proposer/profile picture enrichment.
    - Activity categories: many-to-many join table, idempotent add/remove, batch fetching, and trip-level category listing.
    - Automatic trip-scoped category creation and transactional default-category seeding on trip creation.
  - Improvements
    - Repository layer: Activity, Category, ActivityCategory repos with cursor pagination and batch category fetch to avoid N+1.
    - Service layer: ActivityService and CategoryService with membership checks, transactional flows, auto-create categories, and response assembly.
    - Controller & routing: ActivityController and CategoryController, route wiring under /api/v1/trips/:tripID with TripMemberRequired middleware, input validation and error mapping.
    - Validation/utilities: validators.ExtractUserID; public enum rename for comment entity types (Activity → ActivityEntity, Pitch → PitchEntity).
  - Infra / Migrations
    - New DB migrations: categories (composite PK trip_id+name), activities (dates JSONB, indexes), activity_categories (join table), drop legacy category_name, add unique/composite FK constraints to enforce trip consistency.

- Sequence diagram (activity creation + category linking)
  - Client -> ActivityController: POST /trips/{tripID}/activities (CreateActivityRequest)
  - ActivityController -> ActivityService: CreateActivity(req, userID)
  - ActivityService -> Repository (tx): validate trip & membership; create missing categories (CategoryRepo); insert activity (ActivityRepo)
  - ActivityService -> Repository (tx): AddCategoriesToActivity (ActivityCategoryRepo)
  - Repository (tx) -> DB: commit
  - ActivityService -> ActivityController: ActivityAPIResponse
  - ActivityController -> Client: 201 Created (ActivityAPIResponse)

- Author contribution table (added / removed lines)
| Component | Added Lines | Removed Lines | Files Changed |
|-----------|------------:|-------------:|---------------:|
| Documentation (docs.go, swagger.json, swagger.yaml) | 1,714 | 4 | 3 |
| Controllers | 452 | 0 | 3 |
| Repositories | 433 | 0 | 4 |
| Services | 551 | 0 | 3 |
| Models | 308 | 0 | 4 |
| Database Migrations | 110 | 0 | 7 |
| Routers | 67 | 0 | 3 |
| Tests | 925 | 0 | 8 |
| Updates to Existing Files | 23 | 9 | 5 |
| **Total** | **4,583** | **13** | **40**

### CODE REVIEW

File: backend/internal/migrations/20260211031500_create_activity_categories_table.sql [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+-- +goose StatementBegin
+CREATE TABLE activity_categories (
+    activity_id UUID NOT NULL,
+    trip_id UUID NOT NULL,
+    category_name VARCHAR(255) NOT NULL,
+    created_at TIMESTAMP WITH TIME ZONE DEFAULT now(),
+    PRIMARY KEY (activity_id, category_name),
+    FOREIGN KEY (activity_id) REFERENCES activities(id) ON DELETE CASCADE,
+    FOREIGN KEY (trip_id, category_name) REFERENCES categories(trip_id, name) ON DELETE CASCADE
+);
```
[the remark this answers is not in the document]
Comment (hill-ol) [reply backend/internal/migrations/20260211031500_create_activity_categories_table.sql]: Added a unique constraint on activities(id, trip_id) and then created a composite foreign key from activity_categories(activity_id, trip_id) to activities(id, trip_id). Enforces at the database level that the trip_id in activity_categories must always match the trip_id of the referenced activity.

File: backend/internal/services/activity.go [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+			Exec(ctx)
+		if err != nil {
+			return err
+		}
+		createdActivity = activity
+
+		// If categories provided, create them and link to activity
+		if len(req.CategoryNames) > 0 {
+			// Create categories (idempotent - won't fail on duplicates)
+			for _, categoryName := range req.CategoryNames {
```
Comment (in-mai-space) [inline backend/internal/services/activity.go]: maybe we can insert this in bulk, instead of doing multiple insertion here

**Review by aahiltn — COMMENTED** [review]
WOWOWOW BANGER PR! A few followup questions, but looks pretty good to me! Thanks for taking on a few of those design decisions as well, they make sense and show impressive work :)
